projects/cm_mcu/AlarmUtilities.c

```diff
@@ -139,3 +138,109 @@ struct GenericAlarmParams_t tempAlarmTask = {
     .errorlog_registererror = &TempErrorLog,
     .errorlog_clearerror = &TempClearErrorLog,
 };
+
+///////////////////////////////////////////////////////////
+//
+// Voltage Alarms
+//
+///////////////////////////////////////////////////////////
+
+#define INITIAL_ALARM_VOLT_DCDC 70.0f // FIXME : copy from temp
+#define INITIAL_ALARM_VOLT_TM4C 70.0f // FIXME : copy from temp
+#define ALM_OVERVOLT_THRESHOLD  5.0f  // FIXME : copy from temp
+// current value of the thresholds
+static float alarmVolt[4] = {0.f, INITIAL_ALARM_VOLT_DCDC,
+                             INITIAL_ALARM_VOLT_TM4C, 0.f};
+// current value of voltages
+static float currentVolt[4] = {0.f, 0.f, 0.f, 0.f}; // FIXME : copy from temp
+
+// Status flags of the voltage alarm task
+static uint32_t status_V = 0x0;
+
+// read-only, so no need to use queue
+uint32_t getVoltAlarmStatus()
+{
+  return status_V;
+}
+
+float getAlarmVoltage(enum device theDevice)
+{
+  return alarmVolt[theDevice];
+}
+void setAlarmVoltage(enum device theDevice, float voltage)
+{
+  alarmVolt[theDevice] = voltage;
+}
+
+// check the current voltage status.
+// returns +1 for warning, +2 or higher for error
+int VoltStatus()
+{
+  int retval = 0;
+  status_V = 0x0U;
+
+  // microcontroller
+  currentVolt[TM4C] = getADCvalue(ADC_INFO_VCC_INIT_CH); // start with the adc value from the first voltage channel
+  for (int ch = ADC_INFO_VCC_INIT_CH + 1; ch < ADC_INFO_VCC_FIN_CH + 1; ++ch) {
+    if (getADCvalue(ch) > currentVolt[TM4C])
+      currentVolt[TM4C] = getADCvalue(ch); // find the maximum among voltage channels
+  }
+
+  float excess_volt = currentVolt[TM4C] - getAlarmVoltage(TM4C);
+  if (excess_volt > 0.f) { // over temperature
+    status_V |= ALM_STAT_TM4C_OVERVOLT;
+    retval++;
+    if (excess_volt > ALM_OVERVOLT_THRESHOLD)
+      ++retval;
+  }
+
+  // DCDC. The first command is READ_VOUT.
+  // I am assuming it stays that way!!!!!!!!
+  currentVolt[DCDC] = -99.0f; // FIXME : copy from temp
+  for (int ps = 0; ps < dcdc_args.n_devices; ++ps) {
+    for (int page = 0; page < dcdc_args.n_pages; ++page) {
+      size_t index =
+          ps * (dcdc_args.n_commands * dcdc_args.n_pages) + page * dcdc_args.n_commands + 3;
+      float thisvolt = dcdc_args.pm_values[index];
+      if (thisvolt > currentVolt[DCDC])
+        currentVolt[DCDC] = thisvolt;
+    }
+  }
+  excess_volt = currentVolt[DCDC] - getAlarmVoltage(DCDC);
+  if (excess_volt > 0.f) {
+    status_V |= ALM_STAT_DCDC_OVERVOLT;
+    retval++;
+    if (excess_volt > ALM_OVERVOLT_THRESHOLD)
+      ++retval;
+  }
+
+  return retval;
+}
+
+void VoltErrorLog()
+{
+  log_warn(LOG_ALM, "Voltage high: status: 0x%04x MCU: %d PS:%d\r\n",
+           status_V, (int)currentVolt[TM4C], (int)currentVolt[DCDC]);
+  // errbuffer_volt_high((uint8_t)currentVolt[TM4C],(uint8_t)currentVolt[DCDC]); //FIXME : necessary?
```

Comment:
I think we do want something in the error logger / eeprom as well

Reply:
I guess based on your comment in another thread. I should add error code for not normal voltage [here](https://github.com/apollo-lhc/cm_mcu/blob/dev_alarm_tasks/common/utils.c#L136)? such as "Volt. Normal." and may be "Volt High (TM4C DCDC)" for `EBUF_VOLT_HIGH`?

Reply:
Yes, but we need to think about which values to store and how. It is a bit hard to store all the values at the time.